File: services/rooms/src/app_types/map.py
from enum import StrEnum
from typing import Literal, NamedTuple, NotRequired, TypedDict


class CellType(StrEnum):
    SPAWN = "spawn"
    KING = "king"
    BLOCKER = "block"
    FIELD = "field"
    CASTLE = "castle"


class Cell(TypedDict):
    type: NotRequired[CellType]
    player: NotRequired[int]
    power: NotRequired[int]


class Point(NamedTuple):
    row: int
    col: int


class MapMeta(TypedDict):
    points_of_interest: dict[CellType, list[Point]]
    version: Literal[1]


GameMap = list[list[Cell]]


class MapAndMeta(TypedDict):
    map: GameMap
    meta: MapMeta

File: services/front/src/types/map.ts
export enum CellType {
  SPAWN = "spawn",
  HIDE = "hide",
  KING = "king",
  BLOCKER = "block",
  FIELD = "field",
  CASTLE = "castle",
}

export type Cell = {
  type: CellType;
  power?: number;
  player?: number;
};

export type GameMap = Cell[][];

export type Point = [number, number];

export interface MapMeta {
  points_of_interest: { [key in CellType]: Point[] };
}

export type PlayerData = {
  id: number;
  username: string;
  color: number;
  status: 'ready' | 'win' | 'lose';
}

export type GameStat = {
  fields: number;
  power: number;
}
